icu4c/source/test/cintltst/cloctst.c

```diff
@@ -6665,6 +6667,155 @@ static void test_unicode_define(const char *namech, char ch, const char *nameu,
+              "change memory in the end of buffer while it should. "
+              "Returns %s\n",
+              locale, inLocale, cbuff);
+  }
+  if (buff[length] != 0x0000) {
+      log_err("uloc_getDisplayName(%s, %s) w/ maxResultSize=length+1 chould "
+              "null terminate at buff[length] but does not %x. Returns %s\n",
+              locale, inLocale, buff[length], cbuff);
+  }
+
+  // 2. Test when we only set the maxResultSize to preflight_length.
+
+  // Set sentinal1 in the buff[preflight_length-1] to check it will be replaced
+  // with display name.
+  buff[preflight_length-1] = sentinal1;
+  // Set sentinal2 in the buff[preflight_length] to check it won't be replaced
+  // by null.
+  buff[preflight_length] = sentinal2;
+  status = U_ZERO_ERROR;
+  length = uloc_getDisplayName(locale, inLocale, buff, preflight_length, &status);
+
+  convert_status = U_ZERO_ERROR;
+  // Convert the actual result to UTF8 for debugging output.
+  u_strToUTF8(cbuff, 256, &outlength, buff, length, &convert_status);
+
+  if (length != preflight_length) {
+      log_err("uloc_getDisplayName(%s, %s) w/ maxResultSize=length return "
+              "length %d different from preflight length %d. Returns '%s'\n",
+              locale, inLocale, length, preflight_length, cbuff);
+  }
+  if (U_STRING_NOT_TERMINATED_WARNING != status) {
+      log_err("uloc_getDisplayName(%s, %s) w/ maxResultSize=length should set "
+              "status to U_STRING_NOT_TERMINATED_WARNING but got %d %s. "
+              "Returns %s\n",
+              locale, inLocale, status, myErrorName(status), cbuff);
+  }
+  if (buff[length-1] == sentinal1) {
+      log_err("uloc_getDisplayName(%s, %s) w/ maxResultSize=length does not "
+              "change memory in the end of buffer while it should. Returns "
+              "'%s'\n",
+              locale, inLocale, cbuff);
+  }
+  if (buff[length] != sentinal2) {
+      log_err("uloc_getDisplayName(%s, %s) w/ maxResultSize=length change "
+              "memory beyond maxResultSize to %x. Returns '%s'\n",
+              locale, inLocale, buff[length], cbuff);
+  }
+  if (buff[preflight_length - 1] == 0x0000) {
+      log_err("uloc_getDisplayName(%s, %s) w/ maxResultSize=length null "
+              "terminated while it should not. Return '%s'\n",
+              cbuff, locale, inLocale);
+  }
+
+  // 3. Test when we only set the maxResultSize to preflight_length-1.
+  // Set sentinal1 in the buff[preflight_length-1] to check it will not be
+  // replaced with display name.
+  buff[preflight_length-1] = sentinal1;
+  // Set sentinal2 in the buff[preflight_length] to check it won't be replaced
+  // by null.
+  buff[preflight_length] = sentinal2;
+  status = U_ZERO_ERROR;
+  length = uloc_getDisplayName(
+      locale, inLocale, buff, preflight_length - 1, &status);
+
+  convert_status = U_ZERO_ERROR;
+  // Convert the actual result to UTF8 for debugging output.
```

Comment:
When you get an error code, then the output buffer might not contain anything useful, so don't try to read it.